File: lang/java/src/main/java/ru/dz/mqtt_udp/util/GlobalErrorHandler.java
package ru.dz.mqtt_udp.util;

import java.net.DatagramSocket;
import java.net.SocketException;
import java.util.function.BiConsumer;


/**
 * Singleton for MQTT/UDP error callback.
 */
public class GlobalErrorHandler {

	private static volatile GlobalErrorHandler instance;
	
	private BiConsumer<ErrorType, String> handler = null;

	/**
	 * Singleton for MQTT/UDP error callback.
	 * @return handler instance
	 */
	public static GlobalErrorHandler get()  
	{
		GlobalErrorHandler localInstance = instance;
		if (localInstance == null) {
			synchronized (GlobalErrorHandler.class) {
				localInstance = instance;
				if (localInstance == null) {
					instance = localInstance = new GlobalErrorHandler();
				}
			}
		}
		return localInstance;
	}


	public static BiConsumer<ErrorType, String> getHandler()
	{
		return get().handler;
	}

	public static void setHandler(BiConsumer<ErrorType, String> h)
	{
		get().handler = h;
	}

	
	public static void handleError(ErrorType type, Throwable th)
	{
		handleError(type, th.toString());
	}
	
	public static void handleError(ErrorType type, String description)
	{
		BiConsumer<ErrorType, String> h = get().handler;
		if( h != null )
		{
			h.accept(type, description);
			return;
		}
		
		// TODO logger
		System.out.println(String.format( "MQTT/UDP %s error: %s", type.toString(), description ));
	}

}

File: lang/kotlin/mqtt_udp/src/main/kotlin/dz/ru/mqtt_udp/kt/GlobalErrorHandler.kt
package dz.ru.mqtt_udp.kt

import dz.ru.mqtt_udp.kt.util.ErrorType
import java.util.function.BiConsumer




object GlobalErrorHandler {
    private var handler: BiConsumer<ErrorType?, String?>? = null

    fun getHandler(): BiConsumer<ErrorType?, String?>? {
        return handler
    }

    fun setHandler(h: BiConsumer<ErrorType?, String?>) {
        handler = h
    }


    fun handleError(type: ErrorType, th: Throwable) {
        handleError(type, th.toString())
    }

    fun handleError(type: ErrorType, description: String) {
        val h = handler;
        if(h != null)
            h.accept(type, description)
        else {
            // TODO logger
            //System.out.println(String.format( "MQTT/UDP %s error: %s", type.toString(), description ));
            System.err.println(java.lang.String.format("MQTT/UDP %s error: %s", type.toString(), description))
        }
    }

}